src/server/req_manager.c

```diff
@@ -810,64 +813,73 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
 	if (plist == (struct svrattrl *)0)
 		return (PBSE_NONE);
 
+	/*
+	 * We have multiple attribute lists in play here.  pre_copy is used
+	 * to copy the attributes into pattr prior to calling the action functions.
+	 * This means the object passed to the action function is fully up to date.
+	 * new is attribute list which the action functions are called with.  new
+	 * may be modified by the action functions, so we need to copy these to pattr
+	 * a second time.  Lastly we have attr_save.  It is a copy of the object's
+	 * attributes prior to any change.  If any action function fails, we copy
+	 * attr_copy back to the real attributes before leaving the function.
+	 */
+
 	new = (attribute *)calloc((unsigned int)limit, sizeof(attribute));
 	if (new == (attribute *)0)
 		return (PBSE_SYSTEM);
-	/*
-	 * decode the new attribute values which are in the request,
-	 * copy the corresponding current attribute into a holding array
-	 * and update it with the newly decoded value
-	 */
 
-	/* Below says if 'allow_unkresc' is TRUE, then set 'unkn' param */
-	/* of attr_atomic_set() to '1'; otherwise, set it to '-1' meaning */
-	/* not to allow unknown resource. */
-	if ((rc =attr_atomic_set(plist, pattr, new, pdef, limit, (allow_unkresc?1:-1), privil, bad)) != 0) {
+	/* Below says if 'allow_unkresc' is TRUE, then set 'unkn' param
+	 * of attr_atomic_set() to '1'; otherwise, set it to '-1' meaning
+	 * not to allow unknown resource
+	 */
+	if ((rc = attr_atomic_set(plist, pattr, new, pdef, limit, (allow_unkresc?1:-1), privil, bad)) != 0) {
 		attr_atomic_kill(new, pdef, limit);
 		return (rc);
 	}
 
+	pre_copy = (attribute *)calloc((unsigned int)limit, sizeof(attribute));
```

Comment:
Once you copy the values from one attribute array to another, all resource and list types are no longer there.  You've done a shallow copy from one to the other.  If we were apply new to the object, we couldn't call the action functions with new since all resources would no longer be there.  This is what pre_copy is for.
We do the following:
1) copy pre_copy to pattr (the object's real attributes)
2) call the action functions with new
3) If any action function fails, set pattr back to attr_save
4) If the action functions are successful, we need to copy new to pattr since the action functions might have changed new.

There's a comment there trying to explain this.  If you have any suggestions on how to make it better, please help.